File: 2015/01_Not_Quite_Lisp/part_2/main.kt
/**
 * Author: happ_2h https://github.com/happ-2h
 *
 * Solution for Advent of Code 2015 Day 1.2
 */

import java.io.File

fun main(args: Array<String>) {
  var floor = 0 // Start at floor 0
  var pos   = 0 // Current position


  File("../input.txt").readLines()[0]
    .forEach outer@ {
      // We've hit the basement
      if (floor == -1) return@outer

      /*
        '(' = ascend floor
        ')' = descend floor
       */
      floor = if (it == '(') floor + 1 else floor - 1
      ++pos
    }

  println(pos)
}

File: 2015/01_Not_Quite_Lisp/part_2/main.c
/**
 * Author: happ_2h https://github.com/happ-2h
 *
 * Solution for Advent of Code 2015 Day 1.2
 */

#include <stdio.h>

#define BUF_LEN 0x2000

int main(int argc, char const *argv[]) {
  FILE* ifile = fopen("../input.txt", "r");

  if (ifile) {
    int floor = 0; // Start at floor 0
    int pos   = 0; // Current position

    char ch = ' ';
    while((ch = fgetc(ifile)) != EOF) {
      /*
        '(' = ascend floor
        ')' = descend floor
      */
      floor = ch == '(' ? floor + 1 : floor - 1;

      ++pos;

      // We've hit the basement
      if (floor < 0) break;
    }

    fprintf(stdout, "%d\n", pos);

    fclose(ifile);
    ifile = NULL;
  }
  else {
    perror("Failed to read input.txt");
    return 1;
  }

  return 0;
}